mlflow/langchain/databricks_dependencies.py

```diff
@@ -48,6 +54,54 @@ def _get_vectorstore_from_retriever(retriever) -> Generator[Resource, None, None
             yield DatabricksServingEndpoint(endpoint_name=embeddings.endpoint)
 
 
+def _extract_databricks_dependencies_from_uc_function_toolkit(
+    agent,
+) -> Generator[Resource, None, None]:
+    # Tools are passed into a Langchain as Seq[BaseTool] part of an AgentExecutor
+    # This function looks for an AgentExecutor, extracts all the tools generated from
+    # UC Function Toolkit. From each of these tools it then extracts the Databricks
+    # SQL Warehouse ID and UC Function Names and adds them to resources.
+    import inspect
+
+    from langchain.agents import AgentExecutor
+    from langchain_community.tools import BaseTool
+    from langchain_community.tools.databricks import UCFunctionToolkit
+
+    if isinstance(agent, AgentExecutor):
+        tools = agent.tools
+        warehouse_ids = set()
+        for tool in tools:
+            if isinstance(tool, BaseTool):
+                # Tools here are a part of the BaseTool and have no attribute of a WarehouseID
+                # Extract the global variables of the function defined in the tool to get
+                # the UCFunctionToolkit Constants
+                nonlocal_vars = inspect.getclosurevars(tool.func).nonlocals
+                if "self" in nonlocal_vars and isinstance(
+                    nonlocal_vars.get("self"), UCFunctionToolkit
+                ):
+                    uc_function_toolkit = nonlocal_vars.get("self")
+                    # As we are iterating through each tool, adding a warehouse id everytime is a
+                    # duplicative resouce. Use a set to dedup warehouse ids and add them in the end
+                    warehouse_ids.add(uc_function_toolkit.warehouse_id)
+
+                    # In langchain the names of the tools are modified to have underscores:
+                    # main.catalog.test_func -> main_catalog_test_func
+                    # The original name of the tool is stored as the key in the tools dictionary
+                    # This code finds the correct tool and extract the key
+                    langchain_tool_name = tool.name
+                    filtered_tool_names = [
+                        tool_name
+                        for tool_name, uc_tool in uc_function_toolkit.tools.items()
```

Comment:
To the uc function toolkit we add the tool to the map here where the key is the proper syntax of catalog.schema.name: https://github.com/langchain-ai/langchain/blob/e01c6789c47842475cd074f87a9263674ec37d58/libs/community/langchain_community/tools/databricks/tool.py#L172

So extracting that from here. 

I am not just splitting by '_' because catalog, schema and name have the '_' character and we dont know where to split